gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,237 @@
 #include <string.h>
 
 #include <algorithm>
+#include <sstream>
 #include <string>
 #include <utility>
 #include <vector>
 
 namespace gapir {
 
 std::unique_ptr<InMemoryResourceCache> InMemoryResourceCache::create(
-    void* buffer) {
+    std::shared_ptr<MemoryAllocator> allocator, size_t memoryLimit) {
   return std::unique_ptr<InMemoryResourceCache>(
-      new InMemoryResourceCache(buffer));
+      new InMemoryResourceCache(allocator, memoryLimit));
 }
 
-InMemoryResourceCache::InMemoryResourceCache(void* buffer)
-    : mHead(new Block(0, 0)),
-      mBuffer(static_cast<uint8_t*>(buffer)),
-      mBufferSize(0) {}
+InMemoryResourceCache::InMemoryResourceCache(
+    std::shared_ptr<MemoryAllocator> allocator, size_t memoryLimit)
+    : mAllocator(allocator),
+      mResourceIndex(),
+      mResources(),
+      mMemoryLimit(memoryLimit),
+      mMemoryUse(0),
+      mIDGenerator(0) {}
 
-InMemoryResourceCache::~InMemoryResourceCache() {
-  while (mHead->next != mHead) {
-    destroy(mHead->next);
+InMemoryResourceCache::~InMemoryResourceCache() {}
+
+bool InMemoryResourceCache::putCache(const Resource& res, const void* resData) {
+  if (res.getSize() > mMemoryLimit) {
+    return false;
   }
-  delete mHead;
-}
 
-void InMemoryResourceCache::clear() {
-  mCache.clear();
-  while (mHead->next != mHead) {
-    mHead = destroy(mHead);
+  // If we need to evict anything to get this new entry to fit. Now's the time
+  // to do it.
+  resize(mMemoryLimit - res.getSize());
+
+  // Get a new ID for this cache entry that is larger than any current ID.
+  auto newID = mIDGenerator++;
+
+  // Try to allocate some memory. If we get an allocation faulture, throw more
```

Comment:
I am not entirely sure this is still true, (or at least accurate). The alllocation may fail if there is enough room, but due to fragmentation rather than other applications (since we pre-reserve the entire range).